File: Python/Problems/Problems/problem1208.py
import heapq
import math
import sys

import numpy as np

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem1208(Problem):
    def test(self) -> bool:
        s = "iktqzyazth"
        t = "havakbjzzc"
        maxCost = 78

        expected = 8

        result = self.equalSubstring(s, t, maxCost)

        print(result)

        return result == expected

    def equalSubstring(self, s: str, t: str, maxCost: int) -> int:
        n = len(s)
        currCost = 0
        r = 0
        result = 0
        for i in range(n):
            if i > 0:
                currCost -= abs(ord(s[i - 1]) - ord(t[i - 1]))
            while r < n and currCost <= maxCost:
                currCost += abs(ord(s[r]) - ord(t[r]))
                r += 1
            result = max(result, r - i - 1)
            if r == n:
                if currCost <= maxCost:
                    result = max(result, r - i)
                break

        return result

File: C++/Problems/Problems/problem_1208.cpp
#include "problem_1208.h"

bool problem_1208::test() {
    std::string s = "iktqzyazth";
    std::string t = "havakbjzzc";
    int maxCost = 78;

    int expected = 8;

    auto result = equalSubstring(s, t, maxCost);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_1208::equalSubstring(std::string s, std::string t, int maxCost) {
    int n = s.size();
    int currCost = 0;
    int r = 0;
    int result = 0;
    for (int i = 0; i < n; ++i)
    {
        if (i > 0)
            currCost -= std::abs(s[i - 1] - t[i - 1]);
        while (r < n && currCost <= maxCost)
        {
            currCost += std::abs(s[r] - t[r]);
            r++;
        }
        result = std::max(result, r - i - 1);
        if (r == n)
        {
            if (currCost <= maxCost)
                result = std::max(result, r - i);
            break;
        }
    }

    return result;
}
